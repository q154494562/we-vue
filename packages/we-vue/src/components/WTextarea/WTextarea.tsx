import Vue from 'vue'
import '@/scss/textarea.scss'

export default Vue.extend({
  name: 'w-textarea',

  props: {
    placeholder: String,
    showCounter: {
      type: Boolean,
      default: true,
    },
    rows: {
      type: [Number, String],
      default: 3,
    },
    maxLength: {
      type: [Number, String],
      default: 100,
    },
    disabled: Boolean,
    readonly: Boolean,
    value: {
      type: String,
      default: '',
    },
  },

  data () {
    return {
      lazyValue: this.value as string,
    }
  },

  computed: {
    length (): number {
      return this.lazyValue ? this.lazyValue.length : 0
    },

    internalValue: {
      get (): string {
        return this.lazyValue
      },
      set (val: any) {
        this.lazyValue = val
        this.$emit('input', val)
      },
    },
  },

  watch: {
    value (val) {
      // 有最大字数限制时对超出限制部分进行截取
      if (this.maxLength && val.length > this.maxLength) {
        val = val.slice(0, this.maxLength)
      }

      this.lazyValue = val
    },
  },

  mounted () {
    if (this.maxLength && this.length > this.maxLength) {
      this.lazyValue = this.value.slice(0, this.maxLength as number)
    } else {
      this.lazyValue = this.value
    }
  },

  methods: {
    onFocus (): void {
      this.$emit('focus')
    },

    onBlur (): void {
      this.$emit('blur')
    },
  },

  render () {
    return (
      <div class="weui-cell">
        <div class="weui-cell__bd">
          <textarea
            onChange={this.$emit('change', this.internalValue)}
            onFocus={this.onFocus}
            onBlur={this.onBlur}
            class="weui-textarea"
            ref="rextarea"
            placeholder={this.placeholder}
            rows={this.rows}
            disabled={this.disabled}
            readonly={this.readonly}
            vModel={this.internalValue}
          />
          {
            this.showCounter &&
              <div class="weui-textarea-counter">
                { `${this.length}/${this.maxLength}` }
              </div>
          }
        </div>
      </div>
    )
  },
})
