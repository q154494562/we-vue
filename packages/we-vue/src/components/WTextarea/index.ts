import WTextarea from './WTextarea'

export { WTextarea }
export default WTextarea
