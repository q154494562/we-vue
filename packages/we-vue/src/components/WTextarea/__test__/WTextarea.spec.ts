import { shallowMount } from '@vue/test-utils'
import WTextarea from '../'

describe('textarea', () => {
  test('create', () => {
    const wrapper = shallowMount(WTextarea, {
      propsData: {},
    })

    expect(wrapper.name()).toBe('w-textarea')
    expect(wrapper.classes()).toContain('weui-cell')

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('show counter', () => {
    const wrapper = shallowMount(WTextarea, {
      propsData: {
        showCounter: true,
      },
    })

    expect(wrapper.contains('.weui-textarea-counter')).toBeTruthy()
    expect(wrapper.find('.weui-textarea-counter').text()).toEqual(`${wrapper.vm.length}/${wrapper.vm.maxLength}`)

    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setProps({
      showCounter: false,
    })
    expect(wrapper.contains('.weui-textarea-counter')).toBeFalsy()

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('watch currentValue', () => {
    const wrapper = shallowMount(WTextarea, {
      propsData: {},
    })

    wrapper.setData({
      internalValue: 'current-value',
    })

    expect(wrapper.emitted().input).toBeTruthy()
  })

  test('watch value', () => {
    const wrapper = shallowMount(WTextarea, {
      propsData: {
        maxLength: 10,
      },
    })

    wrapper.setProps({
      value: 'new-value',
    })

    expect(wrapper.vm.internalValue).toBe('new-value')

    // set a value longer than maxLength
    wrapper.setProps({
      value: 'new-value-longer',
    })

    expect(wrapper.vm.internalValue).toBe('new-value-')
  })

  test('check max-length', () => {
    const wrapper = shallowMount(WTextarea, {
      propsData: {
        maxLength: 2,
        value: 'test',
      },
    })

    expect(wrapper.vm.internalValue).toHaveLength(2)
    expect(wrapper.vm.internalValue).toBe('te')
  })

  test('focus event', () => {
    const wrapper = shallowMount(WTextarea, {
      propsData: {},
    })

    wrapper.find('textarea').trigger('focus')

    expect(wrapper.emitted().focus).toBeTruthy()
  })

  test('blur event', () => {
    const wrapper = shallowMount(WTextarea, {
      propsData: {},
    })

    wrapper.find('textarea').trigger('blur')

    expect(wrapper.emitted().blur).toBeTruthy()
  })
})
