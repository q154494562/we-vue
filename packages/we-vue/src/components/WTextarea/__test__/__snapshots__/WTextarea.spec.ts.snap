// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`textarea create 1`] = `
<div class="weui-cell">
  <div class="weui-cell__bd"><textarea rows="3" class="weui-textarea"></textarea>
    <div class="weui-textarea-counter">0/100</div>
  </div>
</div>
`;

exports[`textarea show counter 1`] = `
<div class="weui-cell">
  <div class="weui-cell__bd"><textarea rows="3" class="weui-textarea"></textarea>
    <div class="weui-textarea-counter">0/100</div>
  </div>
</div>
`;

exports[`textarea show counter 2`] = `
<div class="weui-cell">
  <div class="weui-cell__bd"><textarea rows="3" class="weui-textarea"></textarea></div>
</div>
`;
